Blog post about upgrade improvements (#229)
strimzi/strimzi.github.io: _posts/2021-04-07-upgrade-improvements-and-migration.md

---
layout: post
title:  "Upgrade changes and API versions conversion"
date: 2021-04-07
author: jakub_stejskal
---
Upgrade process of Strimzi and Kafka is not trivial. 
In Strimzi 0.22 we introduced several improvements, which will make the process easier for the users. 

> Moreover, we also introduced new API version `v1beta2` which will replace `v1beta1` because [Kubernetes 1.22 will remove support](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122) for `apiextensions/v1beta1`.

Reviewer: I do not think people will understand this. You might need to explain more ... but that would make it too long for the perex. so maybe we could say here just something like ...

```suggestion
Moreover, we also introduced new API version `v1beta2` which will replace `v1beta1` / `v1alpha1` APIs and make Strimzi ready for Kubernetes 1.22.
```

... and leave the details for the main text?

This blog post will show you how the upgrade process changed and what you should do to keep your cluster working without issues.

<!--more-->

### Upgrade improvements

Legacy upgrade process was a little complicated.

> Users had to have fill Kafka `version`, but also `log.message.format.version` or `inter.broker.protocol.version` before the upgrade.

Reviewer: We basically still recommend users to use this ... so I would not call it legacy. I would probably focus most here on the upgrading through each version and on the fact that you needed these even with operator hub etc.

This makes the upgrades hard to execute fully automatically for example when you use Strimzi installed from [OperatorHub](https://operatorhub.io/operator/strimzi-kafka-operator).
You also needed to upgrade through all versions of Strimzi and Kafka.

> For example if you wanted to upgrade from 0.18 to 0.20, you had to install 0.19 firstly and then do the same steps for 0.20.

Reviewer: ```suggestion
For example if you wanted to upgrade from 0.18 to 0.20, the recommended approach was to install 0.19 firstly and then do the same steps for 0.20.
```

> Now, the process is not too strict.

Reviewer: ```suggestion
Now, the process is a bit more friendly.
```

* detect used Kafka `version`, `log.message.format.version` and `inter.broker.protocol.version` (from CR or compute it from Kafka `version`)
* configure Kafka `version`, `log.message.format.version` and `inter.broker.protocol.version` if it's not specified by user

The main improvement is, that you don't need to upgrade across all Strimzi versions, but you can upgrade from for instance 0.18 directly to 0.22 as it is shown in the diagram bellow.
You can face the following states:
* Supported Kafka `version` and `log.message.format.version` or `inter.broker.protocol.version` is set in CR - Strimzi will keep the values and missing one will se based on `version`.
* Supported Kafka `version` is set, `log.message.format.version` and `inter.broker.protocol.version` are not set in CR - Strimzi will set `log.message.format.version` and `inter.broker.protocol.version` based on `version`
* Supported Kafka `version` is not set in CR - Strimzi will set Kafka `version`, `log.message.format.version` and `inter.broker.protocol.version` based on the default Kafka supported in 0.22.
* Unsupported Kafka `version` is set in CR - Strimzi will set `Kafka` CR status to `NotReady`. 

>   In that case user has to update CR and manually set a proper config.

Reviewer: I think you should go into more details here. Especially with regard to what happens when the values are not set. The important point being that they will be:
* First set to the version from the old Kafka => to make it possible to roll the cluster safely
* Then they will be automatically updated to the new version.

You can also finish it with some _how many restarts to expect_ paragraph.

And TBH, I think that would make more interesting diagram than the one you have ;-)

![Upgrade improvements diagram](/assets/images/posts/2021-04-16-upgrade-improvements.png)

Even with these changes, there are still some disadvantages:
* For upgrade, roll out the new brokers while first using the older `log.message.format.version` or `inter.broker.protocol.version` and only afterwards change to the new versions for message format and inter-broker protocol.
* Downgrade won't be executed when new `log.message.format.version` or `inter.broker.protocol.version` are already used.

> You need to change configuration to use same `log.message.format.version` and `inter.broker.protocol.version` as older Kafka.

Reviewer: ```suggestion
You need to change configuration to use same or older `log.message.format.version` and `inter.broker.protocol.version` as older Kafka.
```

  
You can find further information in our [documentation](https://strimzi.io/docs/operators/0.22.1/full/deploying.html#proc-upgrade-cli-tool-crds-str).

### New API versions

As we stated before, we introduced new API version `v1beta2` for Strimzi resources. 
Every user of Strimzi will have to the migration from old `v1beta1` to `v1beta2`.

> This migration is not necessarily needed right after upgrade to Strimzi 0.22, but has to be done before upgrade to Strimzi 0.23.

Reviewer: ```suggestion
This migration is not necessarily needed right after upgrade to Strimzi 0.22, but has to be done before upgrade to Strimzi 0.23 or later.
```

> The main reason why Strimzi users had to do this migration are preparations of Strimzi to use Kubernetes CRD `apiextensions/v1`.

Reviewer: IMHO here you should put the link from the perex.

Custom resources are edited and controlled using APIs added to Kubernetes by CRDs. 
Put another way, CRDs extend the Kubernetes API to allow the creation of custom resources. 
CRDs are themselves resources within Kubernetes. 
They are installed in a Kubernetes cluster to define the versions of API for the custom resource. 
Each version of the custom resource API can define its own schema for that version. 
Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version

> In 0.21, we supported custom resource version `v1alpha` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.

Reviewer: ```suggestion
In 0.21, we supported custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.
```

> These versions are currently deprecated and will be dropped in the following Strimzi minor release.

Reviewer: ```suggestion
These versions are currently deprecated and will be dropped in the following Strimzi 0.23.
```

The tool can operate in two modes:
* convert your resources to `v1beta2`. It can convert resources in yaml files or resource already deployed into Kubernetes cluster.
* convert CRDs to use `v1beta2` as a stored version and use `apiextensions/v1`.

You can use the tool directly from un-archived directory as it's shown in the example bellow.

```bash
$ bin/api-conversion.sh --help
Usage: bin/api-conversion.sh [-hV] [COMMAND]
Conversion tool for Strimzi Custom Resources
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
Commands:
  help                                     Displays help information about the
                                             specified command
  convert-file, cf                         Convert Custom Resources from YAML
                                             file
  convert-resource, cr, convert-resources  Convert Custom Resources directly in
                                             Kubernetes
  crd-upgrade, crd                         Upgrades the Strimzi CRDs and CRs to
                                             use v1beta2 version

> ```

Reviewer: I would really go for a screenshot here.
